crates/node/src/networking/p2p/pea2pea.rs

```diff
@@ -758,6 +756,122 @@ mod tests {
         assert_eq!(into_receive(tx), recv_tx.0);
     }
 
+    // Test 20 peers that connect each other.
+    #[tokio::test]
+    async fn test_twenty_peers() {
+        start_logger(LevelFilter::DEBUG);
+
+        struct Peer {
+            p2p: P2P,
+            tx_sender: UnboundedSender<Transaction<Validated>>,
+            tx_receiver: UnboundedReceiver<(
+                Transaction<Received>,
+                Option<oneshot::Sender<Result<(), EventProcessError>>>,
+            )>,
+        }
+
+        impl Peer {
+            fn new(
+                tuple: (
+                    P2P,
+                    UnboundedSender<Transaction<Validated>>,
+                    UnboundedReceiver<(
+                        Transaction<Received>,
+                        Option<oneshot::Sender<Result<(), EventProcessError>>>,
+                    )>,
+                ),
+            ) -> Peer {
+                Peer {
+                    p2p: tuple.0,
+                    tx_sender: tuple.1,
+                    tx_receiver: tuple.2,
+                }
+            }
+        }
+
+        tracing::debug!("creating P2P beacon node");
+        let p2p_beacon_node = Peer::new(create_peer(&format!("p2p-beacon")).await);
+
+        tracing::debug!("P2P beacon node starts listening");
+        p2p_beacon_node
+            .p2p
+            .node()
+            .start_listening()
+            .await
+            .expect("p2p_beacon_node start_listening()");
+        tracing::debug!("P2P beacon node is listening");
+
+        let num_nodes = 20;
+        let mut peers = Vec::with_capacity(num_nodes);
+
+        for i in 1..num_nodes {
+            tracing::debug!("creating peer {i}");
+            let peer = Peer::new(create_peer(&format!("peer{i}")).await);
+
+            tracing::debug!("peer{i} starts listening");
+            peer.p2p
+                .node()
+                .start_listening()
+                .await
+                .expect("peer{i] listen");
+            tracing::debug!("peer{i} is listening");
+
+            tracing::debug!("peer{i} connects to P2P beacon node");
+            peer.p2p
+                .node()
+                .connect(
+                    p2p_beacon_node
+                        .p2p
+                        .node()
+                        .listening_addr()
+                        .expect("p2p_beacon_node listening_addr()"),
+                )
+                .await
+                .expect(&format!("peer{i} connect p2p_beacon_node"));
+            tracing::debug!("peer{i} is connected to P2P beacon node");
+
+            peers.push(peer);
+        }
+
+        // Let the dust settle down a bit.
+        //tokio::time::sleep(std::time::Duration::from_secs(5)).await;
```

Comment:
@musitdev in the current setting, if there's no wait / sleep between connects and transaction sends, this test locks up on my laptop. What happens is that those derivative new node connects are happening concurrently with the transaction sends and this causes the deadlock somehow.

If you uncomment this sleep and all the connects take place before the transaction sends, so far I couldn't reproduce the deadlock anymore.